app/src/main/java/com/isc/hermes/model/navigation/TrackRecoveryProtocol.java

```diff
@@ -0,0 +1,115 @@
+package com.isc.hermes.model.navigation;
+
+import static com.isc.hermes.model.navigation.NavigationTrackerTools.isPointInsideSegment;
+
+import com.isc.hermes.model.CurrentLocationModel;
+
+import java.util.List;
+
+public class TrackRecoveryProtocol {
+    private final List<RouteSegmentRecord> routeSegments;
+    private final CurrentLocationModel currentLocation;
+    private RouteSegmentRecord foundedTrack;
+
+    public TrackRecoveryProtocol(List<RouteSegmentRecord> routeSegments) {
+        this.routeSegments = routeSegments;
+        currentLocation = CurrentLocationModel.getInstance();
+        foundedTrack = null;
+    }
+
+    public boolean isAttemptSuccessful() {
+        return (foundedTrack != null);
+    }
+
+    public void attemptSimpleRecovery(int routeSegmentIndex) {
+        if (routeSegments.size() == 1) {
+            simpleStrategy();
+        } else if (routeSegments.size() == 2) {
+            doubleWayStrategy(routeSegmentIndex);
+        } else {
+            threeWayStrategy(routeSegmentIndex);
+        }
+
+        System.out.println("Simple recovery is done");
+    }
+
+    public void attemptDeepRecovery() {
+        for (int index = 0; index < routeSegments.size(); index++) {
+            System.out.printf("ATTEMPT RECOVERY ON TRACK #%d\n", index);
+            attemptSimpleRecovery(index);
+            if (isAttemptSuccessful()) {
+                break;
+            } else {
+                System.out.println("ATTEMPT FAILED");
```

Comment:
should implement log statements (Log.i for information and Log.e for error) to provide more meaningful messages.